### PR TITLE
Rework of architecture

### PR DESCRIPTION
**Pull request by eginwong**
1. Use a simple loader of rules for business logic
2. Add configuration to GameState object for future use cases of loading/continuing from a previous state
3. Lots of tests

### CODE REVIEW

File: entities/Rules/index.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    },
+  ],
+  MOVE_PLAYER: [
+    // update position first
+    ({ notify }, gameState) => {
+      gameState.currentPlayer.position += gameState.turnValues.roll.reduce(
+        (acc, val) => acc + val,
+        0
+      );
+      const boardPosition = require("../PropertyService").findBoardPosition(
```
Comment (DivvyByZero) [inline entities/Rules/index.js]: Let's update both the variable name and the function name, here, to something that reflects "currentPlayerBoardPosition" ?

Function: `findPlayerBoardPosition`
Const: `playerBoardPosition`

File: test/Rules/CONTINUE_TURN.test.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+      continueTurnSpy = sinon.spy();
+      rollDiceSpy = sinon.spy();
+      eventBus.on(continueTurnEvent, continueTurnSpy);
+      eventBus.on(rollDiceEvent, rollDiceSpy);
+    });
+
+    it(`should emit desired ${rollDiceEvent} event`, () => {
+      const promptStub = sinon.stub(PlayerActions, "prompt");
+      promptStub.onCall(0).returns("ROLL_DICE");
```
Comment (DivvyByZero) [inline test/Rules/CONTINUE_TURN.test.js]: Shouldn't this also be `rollDiceEvent` ?

File: test/Rules/CONTINUE_TURN.test.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        `${rollDiceEvent} event was not called`
+      );
+    });
+    it("should make a call to the UI#unknownAction if action input is not recognized", () => {
+      const uiSpy = sinon.spy();
+      userInterface.unknownAction = uiSpy;
+
+      const promptStub = sinon.stub(PlayerActions, "prompt");
+      promptStub.onCall(0).returns(undefined);
+      promptStub.onCall(1).returns("ROLL_DICE");
```
Comment (DivvyByZero) [inline test/Rules/CONTINUE_TURN.test.js]: And here - should it not be `rollDiceEvent` ?

File: test/Rules/JAIL.test.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        `Initial UI method for ${inputEvent} was not called`
+      );
+    });
+    it("should set jailed counter", () => {
+      eventBus.emit(inputEvent);
+      expect(gameState.currentPlayer.jailed).to.equal(0, "Jail counter was not set to 0");
+    });
+    it("should update player's position to jail", () => {
+      eventBus.emit(inputEvent);
+      expect(gameState.currentPlayer.position).to.equal(10);
```
Comment (DivvyByZero) [inline test/Rules/JAIL.test.js]: `10` should be coded somewhere / pulled from a mock config / pulled from the config

File: test/Rules/MOVE_PLAYER.test.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+      passGoSpy = sinon.spy();
+      eventBus.on(passGoEvent, passGoSpy);
+    });
+
+    it("should make a call to the UI#playerMovement", () => {
+      gameState.currentPlayer.position = 10;
+      const uiSpy = sinon.spy();
+      userInterface.playerMovement = uiSpy;
+      eventBus.emit(inputEvent);
+      expect(uiSpy.calledOnceWithExactly(13)).to.equal(
```
Comment (DivvyByZero) [inline test/Rules/MOVE_PLAYER.test.js:58]: I would change this to a `const` from the roll values and the old player position as well

Comment (DivvyByZero) [reply test/Rules/MOVE_PLAYER.test.js:58]: Naw, nevermind

File: test/main.test.js [hunk 6]
```diff
@@ -57,33 +43,34 @@ describe("main", () => {
     it(
       gwt`cold start | game is loaded | first player rolls dice and ends turn`,
       () => {
-        require("../entities/Game")(eventBusEmitter, mockUI, gameState);
-
         // arrange
+        const promptStub = sinon.stub(PlayerActions, "prompt");
+        // highest rolling player
```
Comment (DivvyByZero) [inline test/main.test.js:48]: Put this comment beside the below two lines, else add a comment before line 51 with `// player turn` or something ?

Comment (eginwong) [reply test/main.test.js:48]: done!

**Review by DivvyByZero — APPROVED** [review]
Looks good

**Pull request merged**